pkg/auth/handler/api/anonymous_user_promotion_code.go

```diff
@@ -0,0 +1,96 @@
+package api
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/authgear/authgear-server/pkg/api"
+	"github.com/authgear/authgear-server/pkg/api/apierrors"
+	"github.com/authgear/authgear-server/pkg/lib/infra/db/appdb"
+	oauthhandler "github.com/authgear/authgear-server/pkg/lib/oauth/handler"
+	"github.com/authgear/authgear-server/pkg/util/httproute"
+	"github.com/authgear/authgear-server/pkg/util/httputil"
+	"github.com/authgear/authgear-server/pkg/util/log"
+	"github.com/authgear/authgear-server/pkg/util/validation"
+)
+
+func ConfigureAnonymousUserPromotionCodeRoute(route httproute.Route) httproute.Route {
+	return route.
+		WithMethods("POST", "OPTIONS").
+		WithPathPattern("/api/anonymous_user/promotion_code")
+}
+
+var AnonymousUserPromotionCodeAPIRequestSchema = validation.NewSimpleSchema(`
+	{
+		"type": "object",
+		"additionalProperties": false,
+		"properties": {
+			"session_type": { "type": "string" },
+			"refresh_token": { "type": "string" }
+		},
+		"required": ["session_type"]
+	}
+`)
+
+var AnonymousUserPromotionCodeResponseSchema = validation.NewSimpleSchema(`
+	{
+		"type": "object",
+		"properties": {
+			"promotion_code": { "type": "string" },
+			"expire_at": { "type": "string" }
+		},
+		"required": ["promotion_code", "expire_at"]
+	}
+`)
+
+type AnonymousUserPromotionCodeRequest struct {
+	SessionType  oauthhandler.WebSessionType `json:"session_type"`
+	RefreshToken string                      `json:"refresh_token"`
+}
+
+func (p *AnonymousUserPromotionCodeRequest) Validate(ctx *validation.Context) {
+	if !p.SessionType.IsValid() {
+		ctx.Child("session_type").EmitErrorMessage("unknown session type")
```

Comment:
This custom validation can be replaced by using `enum` in the JSON schema directly.